src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexEmbeddedCompletionProvider.cs

```diff
@@ -0,0 +1,560 @@
+        {
+            if (!item.Properties.TryGetValue(StartKey, out var startString) ||
+                !item.Properties.TryGetValue(LengthKey, out var lengthString) ||
+                !item.Properties.TryGetValue(NewTextKey, out var newText))
+            {
+                return SpecializedTasks.Default<CompletionChange>();
+            }
+
+            item.Properties.TryGetValue(NewPositionKey, out var newPositionString);
+
+            return Task.FromResult(CompletionChange.Create(
+                new TextChange(new TextSpan(int.Parse(startString), int.Parse(lengthString)), newText),
+                newPositionString == null ? default(int?) : int.Parse(newPositionString)));
+        }
+
+        public override Task<CompletionDescription> GetDescriptionAsync(Document document, CompletionItem item, CancellationToken cancellationToken)
+        {
+            if (!item.Properties.TryGetValue(DescriptionKey, out var description))
+            {
+                return SpecializedTasks.Default<CompletionDescription>();
+            }
+
+            return Task.FromResult(CompletionDescription.Create(
+                ImmutableArray.Create(new TaggedText(TextTags.Text, description))));
+        }
+    }
+
+    internal class EmbeddedCompletionContext
+    {
+        private readonly RegexEmbeddedCompletionProvider _provider;
+        private readonly CompletionContext _context;
+
+        public readonly List<RegexItem> Items = new List<RegexItem>();
+        public readonly HashSet<string> Names = new HashSet<string>();
+
+        public readonly RegexTree Tree;
+        public readonly SyntaxToken StringToken;
+
+        public EmbeddedCompletionContext(
+            RegexEmbeddedCompletionProvider provider, 
+            CompletionContext context, 
+            RegexTree tree,
+            SyntaxToken stringToken)
+        {
+            _provider = provider;
+            _context = context;
+            Tree = tree;
+            StringToken = stringToken;
+        }
+
+        public int Position => _context.Position;
+        public OptionSet Options => _context.Options;
+        public Document Document => _context.Document;
+        public CompletionTrigger Trigger => _context.Trigger;
+        public CancellationToken CancellationToken => _context.CancellationToken;
+
+        public void AddIfMissing(
+            string displayText, string suffix, string description,
+            RegexNode parentOpt, int? positionOffset = null, string insertionText = null)
+        {
+            var item = _provider.CreateItem(
+                StringToken, displayText, suffix, description, this,
+                parentOpt, positionOffset, insertionText);
+
+            AddIfMissing(item);
+        }
+
+        public void AddIfMissing(RegexItem item)
+        {
+            if (this.Names.Add(item.DisplayText))
+            {
+                this.Items.Add(item);
+            }
+        }
+    }
+
+    internal struct RegexItem
```

Comment:
Nested class if this isn't intended for other systems? Seems specific to completion more or less I guess?

Reply:
Done.